src/main/java/org/opensearch/security/action/apitokens/ApiTokenAction.java

```diff
@@ -295,4 +354,99 @@ private void sendErrorResponse(RestChannel channel, RestStatus status, String er
         }
     }
 
+    /**
+     *  Validates that the user has the required permissions to create an API token (must be a subset of their own permissions)
+     * */
+    @SuppressWarnings("unchecked")
+    void validateUserPermissions(List<String> clusterPermissions, List<ApiToken.IndexPermission> indexPermissions) {
+        final User user = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        final TransportAddress caller = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+        final Set<String> roles = privilegesEvaluator.mapRoles(user, caller);
+
+        // Early return conditions
+        if (roles.isEmpty()) {
+            throw new OpenSearchException("User does not have any roles");
+        } else if (roles.contains("all_access")) {
```

Comment:
Is this assumption safe?

Reply:
Probably not and it's unnecessary so I removed it in: [bf31791](https://github.com/opensearch-project/security/pull/5012/commits/bf317912bcc7ca22df56c2db2d787aedb454a0b3)